### PR TITLE
Clean up evaluation.py

### PR DESCRIPTION
**Pull request by mschwager**
Working on this for https://github.com/returntocorp/semgrep/issues/2548. `evaluation.py` is starting to make my head explode trying to follow everything. This cleans it up a bit :+1:

### CODE REVIEW

File: semgrep/semgrep/evaluation.py [hunk 1]
```diff
@@ -172,129 +172,104 @@ def _evaluate_single_expression(
     flags: Optional[Dict[str, Any]] = None,
 ) -> Set[Range]:
 
-    assert expression.pattern_id, f"<internal error: expected pattern id: {expression}>"
```
Comment (mschwager) [inline semgrep/semgrep/evaluation.py:175]: `assert`s are a no-op most of the time (e.g. when using `-OO` for optimization, which Python does when building wheels). The below `if` check should cover this :+1:

File: semgrep/semgrep/evaluation.py [hunk 2]
```diff
@@ -486,13 +459,11 @@ def _evaluate_expression(
     steps_for_debugging: List[Dict[str, Any]],
     flags: Optional[Dict[str, Any]] = None,
 ) -> Set[Range]:
-    if (
-        expression.operator == OPERATORS.AND_EITHER
-        or expression.operator == OPERATORS.AND_ALL
-    ):
-        assert (
```
Comment (mschwager) [inline semgrep/semgrep/evaluation.py:493]: Same as above, if we care about this `assert` we should perform a real check and `raise` if necessary.

File: semgrep/semgrep/evaluation.py [hunk 3]
```diff
@@ -340,14 +314,6 @@ def group_by_pattern_id(
     return by_id
 
 
-def safe_relative_to(a: Path, b: Path) -> Path:
```
Comment (mschwager) [inline semgrep/semgrep/evaluation.py:343]: This function is unused :broom:

File: semgrep/semgrep/evaluation.py (hunk 1 above)
Comment (brendongo) [reply semgrep/semgrep/evaluation.py:175]: I think the below if will silently continue if the pattern_id is falsy? (which i guess is the same behavior as an optimized out assert) Do we want to make this noisier or is it fine to continue?

File: semgrep/semgrep/evaluation.py [hunk 4]
```diff
@@ -172,129 +171,105 @@ def _evaluate_single_expression(
...
-    metavars_for_patterns = get_metavar_debugging_info(
-        expression, pattern_ids_to_pattern_matches
+    ranges_for_pattern = (
+        [x.range for x in pattern_ids_to_pattern_matches.get(expression.pattern_id, [])]
+        if expression.pattern_id
+        else []
     )
 
     if expression.operator == OPERATORS.AND:
-        add_debugging_info(
```
Comment (brendongo) [inline semgrep/semgrep/evaluation.py:185]: Nice

**Review by brendongo — APPROVED** [review]

File: semgrep/semgrep/evaluation.py (hunk 1 above)
Comment (mschwager) [reply semgrep/semgrep/evaluation.py:175]: I'll convert it to a `raise SemgrepError` to mimic the previous behavior :+1:

**Pull request merged**